src/metrics/include/metrics/operation.hpp

```diff
@@ -123,13 +147,24 @@ class OperationImpl final : private boost::noncopyable {
         return _events;
     }
 
-    void reportAt(time_point finished, OperationEvent<ClockSource>&& event) {
+    void reportAt(time_point started, time_point finished, OperationEvent<ClockSource>&& event) {
+        if (_threshold) {
+            _threshold->counter.total++;
+            if ((finished - started) > _threshold->duration) {
+                _threshold->counter.failed++;
+            }
+
+            if (_threshold->counter.failedPercentage() > _threshold->failedPercentageThreshold) {
+                throw OperationThresholdExceededException();
+            }
+        }
```

Comment:
imho this logic makes more sense as a method on the `OperationThreshold`:

```c++
if (threshold) { threshold->check(finished,started); }
```

This may also make it easier to test the `OperationThreshold` impl.

Reply:
+1